src/meta/processors/admin/CreateSnapshotProcessor.cpp

```diff
@@ -0,0 +1,107 @@
+/* Copyright (c) 2019 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+
+#include "meta/processors/admin/CreateSnapshotProcessor.h"
+#include "meta/processors/admin/SnapShot.h"
+#include "meta/ActiveHostsMan.h"
+
+namespace nebula {
+namespace meta {
+using SignType = storage::cpp2::EngineSignType;
+
+void CreateSnapshotProcessor::process(const cpp2::CreateSnapshotReq& req) {
+    UNUSED(req);
+
+    if (clusterCheck_->isUnblockingRunning()) {
+        LOG(ERROR) << "Write unblocking task is running";
+        resp_.set_code(cpp2::ErrorCode::E_CONFLICT);
+        onFinished();
+        return;
+    }
+
+    auto snapshot = genSnapshotName();
+    folly::SharedMutex::WriteHolder wHolder(LockUtils::snapshotLock());
+
+    auto hosts = ActiveHostsMan::getActiveHosts(kvstore_);
+    if (hosts.empty()) {
+        LOG(ERROR) << "There is no active hosts";
+        resp_.set_code(cpp2::ErrorCode::E_NO_HOSTS);
+        onFinished();
+        return;
+    }
+
+    // step 1 : Let recode the snapshot to meta , default snapshot status is CREATING.
+    //          The purpose of this is to handle the failure of the checkpoint.
+    std::vector<kvstore::KV> data;
+    data.emplace_back(MetaServiceUtils::snapshotKey(snapshot),
+                      MetaServiceUtils::snapshotVal(cpp2::SnapshotStatus::INVALID,
+                                                    NetworkUtils::toHosts(hosts)));
+
+    if (!doSyncPut(std::move(data))) {
+        LOG(ERROR) << "Write snapshot meta error";
+        resp_.set_code(cpp2::ErrorCode::E_STORE_FAILURE);
+        onFinished();
+        return;
+    }
+
+    // step 2 : Blocking all writes action for storage engines.
+    auto signRet = Snapshot::instance(kvstore_)->blockingWrites(SignType::BLOCK_ON);
+    if (signRet != cpp2::ErrorCode::SUCCEEDED) {
+        LOG(ERROR) << "Send blocking sign to storage engine error";
+        resp_.set_code(signRet);
+        clusterCheck_->addUnblockingTask();
+        onFinished();
+        return;
+    }
+
+    // step 4 : Create checkpoint for all storage engines and meta engine.
+    auto csRet = Snapshot::instance(kvstore_)->createSnapshot(snapshot);
+    if (csRet != cpp2::ErrorCode::SUCCEEDED) {
+        LOG(ERROR) << "Checkpoint create error on storage engine";
+        resp_.set_code(csRet);
+        clusterCheck_->addUnblockingTask();
```

Comment:
When creating failed,   you should just unblock writes and return directly. 
The clusterChecker will check the snapshot status at interval,   and rollback the invalid ones.